### PR TITLE
BUG: Fix is_categorical_dtype for Sparse[category]

### LINKED ISSUE #35793: BUG: is_categorical_dtype returns False for Sparse[category, nan]
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
from pandas.api.types import is_categorical_dtype

s = pd.Series(
    ["a", "b", "c", "a", "b", "c"],
    dtype=pd.SparseDtype(pd.CategoricalDtype(['a', 'b', 'c']))
)

print(s.dtype)
# out: Sparse[category, nan]
print(is_categorical_dtype(s))
# out: False
```

#### Problem description
The `is_categorical_dtype` function returns `False` when sparse. This is inconsistent with other types (example for unsigned integer below).

```python
from pandas.api.types import is_unsigned_integer_dtype

s = pd.Series(
    [0, 1, 2, 3],
    name="pd_unsigned_integer",
    dtype=pd.SparseDtype(pd.UInt8Dtype(), pd.NA)
)
print(s.dtype)
# out: Sparse[UInt8, <NA>]
print(is_unsigned_integer_dtype(s))
# out: True
```

```python
s = pd.Series(
    ["a", "b", "c", "a", "b", "c"],
    dtype=pd.CategoricalDtype(['a', 'b', 'c'])
)

print(s.dtype)
# out: category
print(is_categorical_dtype(s))
# out: True
```

#### Expected Output

```python
print(is_categorical_dtype(s))
# out: True
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.1.0
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 45.0.0.post20200113
Cython           : None
pytest           : 5.3.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.1.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.47.0

</details>

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #35793
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/dtypes/common.py [hunk 1]
```diff
@@ -558,8 +558,9 @@ def is_categorical_dtype(arr_or_dtype) -> bool:
     """
     if isinstance(arr_or_dtype, ExtensionDtype):
         # GH#33400 fastpath for dtype object
-        return arr_or_dtype.name == "category"
-
+        return "category" in arr_or_dtype.name
```
Comment (jreback) [inline pandas/core/dtypes/common.py:561]: don't change the ==

File: pandas/core/dtypes/common.py [hunk 2]
```diff
@@ -558,8 +558,9 @@ def is_categorical_dtype(arr_or_dtype) -> bool:
...
+    if hasattr(arr_or_dtype, "dtype"):
```
Comment (jreback) [inline pandas/core/dtypes/common.py:562]: dont recurse; call the method below

**Review by jreback — CHANGES_REQUESTED** [review]
pls perf test - this is a hot spot

**Comment by dsaxton** [comment]
Ran the categoricals.py asvs and somehow performance improved (?):
```
       before           after         ratio
     [97ec0627]       [b52a7eab]
     <master>         <sparse-categorical-dtype>
-      10.0±0.3ms       9.05±0.2ms     0.90  categoricals.Isin.time_isin_categorical('int64')
-      13.5±0.2ms         11.9±1ms     0.88  categoricals.Isin.time_isin_categorical('object')
-        5.75±2μs       4.00±0.1μs     0.69  categoricals.Indexing.time_get_loc

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Comment by dsaxton** [comment]
mypy error seems wrong, dtype.subtype comes after a hasattr(dtype, "subtype") check:
```
pandas/core/dtypes/base.py:292: error: "object" has no attribute "subtype"  [attr-defined]
```

File: pandas/core/dtypes/base.py [hunk 3]
```diff
@@ -289,6 +289,8 @@ def is_dtype(cls, dtype: object) -> bool:
             return False
         elif isinstance(dtype, cls):
             return True
+        elif hasattr(dtype, "subtype") and isinstance(dtype.subtype, cls):
```
Comment (arw2019) [inline pandas/core/dtypes/base.py:292]: @dsaxton I think replacing `dtype.subtype` &#8594;  `getattr(dtype, "subtype", None)` like you said does silence that `mypy` complaint

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Is this change actually something we necessarily want?

For example one use case where you might check that something is a categorical, is to know that you can then use certain APIs specific to categoricals. But if a Sparse type also gives true, that won't longer be true.

**Comment by dsaxton** [comment]
> Is this change actually something we necessarily want?
> 
> For example one use case where you might check that something is a categorical, is to know that you can then use certain APIs specific to categoricals. But if a Sparse type also gives true, that won't longer be true.

Hmm, interesting point. I think the argument from the original issue seems valid though, that if Sparse[int] is an int, then Sparse[categorical] should be a categorical (making something sparse shouldn't change that). I suppose adopting that view would mean that there's really less of an API specific to categoricals, so some special casing might have to happen as a result.

**Comment by jorisvandenbossche** [comment]
Yes, but as I noted on the issue: if there is an inconsistency, there are always two ways this can be solved. Meaning: we can also change Sparse[int] to no longer be int, or at least check what the use cases are for it being considered int (there quite possibly are, but I think we should try to be concrete first before deciding how to solve the inconsistency)

**Comment by dsaxton** [comment]
> Meaning: we can also change Sparse[int] to no longer be int, or at least check what the use cases are for it being considered int (there quite possibly are, but I think we should try to be concrete first before deciding how to solve the inconsistency)

The more I think about it the more this seems like the better option over changing this for categorical; is_categorical_dtype loses most of it's usefulness if it doesn't guarantee access to the API specific to categoricals.

So then to be consistent sparse dtypes should really only register as sparse (not int, float, whatever)? I would be curious if that's actually easier internally, or what that might break if changed (or if you even bother changing it at all).

**Comment by dsaxton** [comment]
@jorisvandenbossche One possible counterpoint is that is_categorical_dtype already doesn't guarantee access to the categorical API:

```python
[ins] In [1]: import pandas as pd

[ins] In [2]: from pandas.core.dtypes.common import is_categorical_dtype

[ins] In [3]: s = pd.Series(pd.Categorical(["a", "b", "c"]))

[ins] In [4]: is_categorical_dtype(s)
Out[4]: True

[ins] In [5]: s.codes
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-c9d8be9c51ce> in <module>
----> 1 s.codes

~/opt/miniconda3/envs/pandas-1.1.2/lib/python3.8/site-packages/pandas/core/generic.py in __getattr__(self, name)
   5134             if self._info_axis._can_hold_identifiers_and_holds_name(name):
   5135                 return self[name]
-> 5136             return object.__getattribute__(self, name)
   5137
   5138     def __setattr__(self, name: str, value) -> None:

AttributeError: 'Series' object has no attribute 'codes'
```

That assumption caused a bug here: https://github.com/pandas-dev/pandas/issues/36383

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -233,7 +233,7 @@ Bug fixes
 Categorical
 ^^^^^^^^^^^
 
--
+- Bug in :func:`pandas.core.dtypes.common.is_categorical_dtype` where sparse categorical dtypes would return ``False`` (:issue:`35793`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:236]: use the public facing api here `pandas.dtypes.is_categorical_dtype`

File: pandas/core/dtypes/common.py [hunk 5]
```diff
@@ -558,8 +558,7 @@ def is_categorical_dtype(arr_or_dtype) -> bool:
     """
     if isinstance(arr_or_dtype, ExtensionDtype):
         # GH#33400 fastpath for dtype object
-        return arr_or_dtype.name == "category"
-
+        return ("category" == arr_or_dtype.name) or ("category" in arr_or_dtype.name)
```
Comment (jreback) [inline pandas/core/dtypes/common.py:561]: i would be explict about this being a sparse type and check the subtype

File: pandas/core/dtypes/base.py [hunk 6]
```diff
@@ -287,6 +287,10 @@ def is_dtype(cls, dtype: object) -> bool:
             return False
         elif isinstance(dtype, cls):
             return True
+        elif hasattr(dtype, "subtype") and isinstance(
```
Comment (jreback) [inline pandas/core/dtypes/base.py:290]: do we *really* need to add this, this is a perf sensitive piece.

File: pandas/tests/dtypes/test_common.py [hunk 7]
```diff
@@ -213,6 +213,15 @@ def test_is_categorical_deprecation():
         com.is_categorical([1, 2, 3])
 
 
+def test_sparse_categorical_is_categorical():
```
Comment (jreback) [inline pandas/tests/dtypes/test_common.py:216]: test_is_categorical_sparse_categorical

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/base.py (hunk 6 above)
Comment (dsaxton) [reply pandas/core/dtypes/base.py:290]: It's needed for test_is_categorical_sparse_categorical to pass; I'm not sure if there's another way to make this work. I ran the dtype asvs and didn't see a significant degradation.

**Comment by jreback** [comment]
code change looks ok. but I think @jorisvandenbossche is right here; if `is_categorical_dtype` is true then this should *act* like a categorical, which we currently don't do.

File: doc/source/whatsnew/v1.2.0.rst [hunk 8]
```diff
@@ -236,7 +236,7 @@ Bug fixes
 Categorical
 ^^^^^^^^^^^
 
--
+- Bug in :func:`pandas.dtypes.is_categorical_dtype` where sparse categorical dtypes would return ``False`` (:issue:`35793`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.2.0.rst:239]:
```suggestion
- Bug in :func:`pandas.api.types.is_categorical_dtype` where sparse categorical dtypes would return ``False`` (:issue:`35793`)
```

The public import is a bit different

**Comment by jorisvandenbossche** [comment]
> One possible counterpoint is that is_categorical_dtype already doesn't guarantee access to the categorical API:

But it does guarantee that on a Series `.cat.codes` is available (it's true you still need to know if it's a Series[Categorical] or Categorical).

Also, do we actually support Sparse[category] properly? 
Eg I just took the example from the test you added, and the repr fails:

```
In [1]: s = pd.Series(["a", "b", "c"], dtype=pd.SparseDtype(pd.CategoricalDtype(["a", "b", "c"])))                                                                                                                 

In [2]: s                                                                                                                                                                                                          
...
ValueError: object __array__ method not producing an array
```

Some time ago I opened a general issue about the question to support Sparse backed by other ExtensionArrays or not: https://github.com/pandas-dev/pandas/issues/26407

**Comment by dsaxton** [comment]
Closing since it seems there isn't a desire to change this

**Pull request closed without merging**